mpmp/utilities/batch_utilities.py

```diff
@@ -5,52 +5,129 @@
     -------
     corrected_data: samples x features dataframe of batch corrected input
+    coefs: coefficients from remove_batch_effect
     """
     if verbose:
         print('Correcting for batch effects using limma...', file=sys.stderr)
 
-    limma = importr('limma')
-
-    # limma expects data in features x samples format
-    if correct_covariates:
-        values_to_correct = data.copy().values.T
+    if columns is not None:
+        values_to_correct = data.loc[:, columns].copy().values
     else:
-        values_to_correct = data.loc[:, gene_features].copy().values.T
-
-    pandas2ri.activate()
-
-    corrected_values = limma.removeBatchEffect(values_to_correct, batches)
+        values_to_correct = data.copy().values
 
+    corrected_values, coefs = remove_batch_effect(values_to_correct,
+                                                  batches,
+                                                  coefs=coefs)
     corrected_data = data.copy()
 
-    if correct_covariates:
-        corrected_data.loc[:, :] = corrected_values.T
+    if columns is not None:
+        corrected_data.loc[:, columns] = corrected_values
     else:
-        corrected_data.loc[:, gene_features] = corrected_values.T
+        corrected_data.loc[:, :] = corrected_values
 
     # TODO: maybe add unit test for this?
     assert corrected_data.columns.equals(data.columns)
     assert corrected_data.shape == data.shape
 
-    return corrected_data
+    return corrected_data, coefs
+
+
+def remove_batch_effect(X, batches, coefs=None):
+    """Python version of limma::removeBatchEffect.
+
+    This should duplicate the original R code here (for the case
+    where there is only a single vector of batches):
+    https://rdrr.io/bioc/limma/src/R/removeBatchEffect.R
+
+    For now, batches needs to be integer indexes.
+
+    If coefs are provided, they should be an m x p vector, where m
+    is the dimension of the design matrix and p is the number of features
+    in the original dataset.
+    """
+    from patsy.contrasts import Sum
+
+    # use sum coding to code batches, this is what limma does
+    # https://www.statsmodels.org/dev/examples/notebooks/generated/contrasts.html#Sum-(Deviation)-Coding
+    # this is a bit easier/more intuitive in R, due to its built-in factor
+    # type, but we can sort of emulate it here with pandas categorical data
+    batches_df = pd.Series(batches, dtype='category')
+    contrast = Sum().code_without_intercept(
+        list(batches_df.cat.categories)
+    )
+    design = contrast.matrix[batches.astype(int), :]
+
+    # if coefficients are provided, just use them to correct the provided data
+    # otherwise fit the model and correct the provided data
+    if coefs is None:
+        from sklearn.linear_model import LinearRegression
+        # X is an n x p matrix
+        # batches is a n x m vector of batch indicators
+        # we want to find a m x p vector of coefficients
+        reg = LinearRegression().fit(design, X)
```

Comment:
Double checking: limma uses the design matrix to predict the data rather than just predicting the class from the labels and subtracting the residuals?

Reply:
That is to say that X is the dependent variable and design is the independent?

Reply:
> That is to say that X is the dependent variable and design is the independent?

Yes, that's right. The idea is to predict the original data matrix as a linear function of the batch information, then subtract away that approximation to get the corrected/adjusted data matrix. This idea is described [in this paper](https://www.biorxiv.org/content/10.1101/2020.08.17.255034v1.full.pdf) (see the "Problems of input adjustment" section), and the R code from limma is [here](https://www.biostars.org/p/434736/) which should do the same thing (although they don't explicitly describe it anywhere as far as I know).

> just predicting the class from the labels and subtracting the residuals

How would this work? It sounds kind of similar to what I'm doing but I'm not sure I totally follow.

Reply:
But then the residuals would be in Y-space, right? Or in other words, there would just be a single residual for the prediction of Y, rather than one for each column in X - would you just subtract the same residual from each column in X?

I guess these are similar ideas, it just makes more intuitive sense to me to have a different prediction/residual/correction for each column of X. I'm not super convinced that it's "right", though, I think there's probably lots of ways to do this.

Reply:
> But then the residuals would be in Y-space, right? Or in other words, there would just be a single residual for the prediction of Y, rather than one for each column in X - would you just subtract the same residual from each column in X?
> 
> I guess these are similar ideas, it just makes more intuitive sense to me to have a different prediction/residual/correction for each column of X. I'm not super convinced that it's "right", though, I think there's probably lots of ways to do this.

After looking up how statistics works, I think it would be roughly the same idea in the other direction. Instead of subtracting out the signal from the data you'd subtract out the linear prediction from the labels. What remains would be the parts of the label unpredictable via linear model.

It's less intuitive, but it works for resnets so should be at least possible (though not necessarily helpful)

Reply:
> you'd subtract out the linear prediction from the labels

Ahh, got it -- so you'd be modifying the labels rather than the data. That does seem like it would have a similar effect (as far as removing linear signal) - maybe worth trying in the future.